pkg/manager/member/scaler.go

```diff
@@ -232,3 +265,53 @@ func normalizeDeleteSlots(replicas int32, deleteSlots sets.Int32, desiredDeleteS
 	}
 	return deleteSlots
 }
+
+// setReplicasAndDeleteSlotsByFinished sets the replica and deleteSlots according to the ordinals finished successfully.
+// parameter means:
+// - scaling:
+//   - 1: scaling out
+//   - -1: scaling in
+// - ordinals: pod oridnals to create or delete this round, in processing order, which means increasing when scale out and decreasing when scale in.
+// - finishedOrdinals: oridnals finished successfully in this round.
+func setReplicasAndDeleteSlotsByFinished(scaling int, newSet, oldSet *apps.StatefulSet, ordinals []int32, finishedOrdinals sets.Int32) {
+	klog.Infof("scale statefulset: ordinals: %v, finishedOrdinals: %v", ordinals, finishedOrdinals)
+	newReplicas := *oldSet.Spec.Replicas
+	actualPodOrdinals := helper.GetPodOrdinals(*oldSet.Spec.Replicas, oldSet)
+	actualDeleteSlots := helper.GetDeleteSlots(oldSet)
+	desiredDeleteSlots := helper.GetDeleteSlots(newSet)
+	// copy delete slots from desired delete slots if not in actual pod ordinals
+	for i := range desiredDeleteSlots {
+		if !actualPodOrdinals.Has(i) {
+			actualDeleteSlots.Insert(i)
+		}
+	}
+	// we should count ordinal by order strict here.
+	if scaling > 0 {
+		for _, ordinal := range ordinals {
+			if !finishedOrdinals.Has(ordinal) {
+				break
+			}
+			newReplicas++
+			if !desiredDeleteSlots.Has(ordinal) {
+				actualDeleteSlots.Delete(ordinal)
+			}
+		}
+	} else {
+		for _, ordinal := range ordinals {
+			if !finishedOrdinals.Has(ordinal) {
+				break
+			}
+			newReplicas--
+			if desiredDeleteSlots.Has(ordinal) {
+				actualDeleteSlots.Insert(ordinal)
+			}
+		}
+	}
+	if newReplicas == *oldSet.Spec.Replicas {
```

Comment:
This is always equal due to L278?

Reply:
We may change newReplicas in L294 and L304.